stackslib/src/chainstate/nakamoto/mod.rs

```diff
@@ -334,108 +365,514 @@ impl NakamotoBlockHeader {
+                    warn!(
+                        "Invalid Stacks block header {}: failed to verify VRF proof: {}",
+                        self.header.block_hash(),
+                        e
+                    );
+                    false
+                }
+            };
+
+            if !valid {
+                warn!("Invalid Nakamoto block: leader VRF key did not produce a valid proof";
+                      "block_id" => %self.block_id(),
+                      "leader_public_key" => %leader_key.public_key.to_hex(),
+                      "sortition_hash" => %burn_chain_tip.sortition_hash
+                );
+                return Err(ChainstateError::InvalidStacksBlock(
+                    "Invalid Nakamoto block: leader VRF key did not produce a valid proof".into(),
+                ));
+            }
+        }
 
-                cur_sortition_tip =
-                    SortitionDB::get_block_snapshot(sortition_db_conn, &parent_sortition_id)?
-                        .ok_or_else(|| ChainstateError::NoSuchBlockError)?
-                        .consensus_hash;
+        // this block must commit to all of the work seen so far
+        if self.header.burn_spent != burn_chain_tip.total_burn {
+            warn!("Invalid Nakamoto block header: invalid total burns";
+                  "header.burn_spent" => self.header.burn_spent,
+                  "burn_chain_tip.total_burn" => burn_chain_tip.total_burn
+            );
+            return Err(ChainstateError::InvalidStacksBlock(
+                "Invalid Nakamoto block: invalid total burns".into(),
+            ));
+        }
+        // not verified by this method:
+        // * chain_length       (need parent block header)
+        // * parent_block_id    (need parent block header)
+        // * block-commit seed  (need parent block)
+        // * tx_merkle_root     (already verified; validated on deserialization)
+        // * state_index_root   (validated on process_block())
+        Ok(())
+    }
 
-                // this is true at the "genesis" sortition: this means that there is no stacks block produced yet.
-                if cur_sortition_tip == FIRST_BURNCHAIN_CONSENSUS_HASH {
-                    return Err(ChainstateError::NoSuchBlockError);
+    /// Static sanity checks on transactions.
+    /// Verifies:
+    /// * the block is non-empty
+    /// * that all txs are unique
+    /// * that all txs use the given network
+    /// * that all txs use the given chain ID
+    /// * if this is a tenure-start tx, that:
+    ///    * it has a well-formed coinbase
+    ///    * all TenureChange transactions are present and in the right order, starting with
+    ///    `stacks_tip` and leading up to this block
+    /// * that only epoch-permitted transactions are present
+    pub fn validate_transactions_static(
+        &self,
+        mainnet: bool,
+        chain_id: u32,
+        epoch_id: StacksEpochId,
+    ) -> bool {
+        if self.txs.len() == 0 {
```

Comment:
This is a nit that will prob get fixed in a clippy PR, but in general when doing len() checks against 0, should use is_empty() (otherwise silence the clippy warning)